nativesoundsystem/src/main/jni/aaudio/AAudioManager.cpp

```diff
@@ -18,74 +18,96 @@
 
 #include "AAudioManager.h"
 
-/*
- * PlayAudioThreadProc()
- *   Rendering audio frames continuously; if user asks to play audio, render
- *   sine wave; if user asks to stop, renders silent audio (all 0s)
- *
- */
-void* PlayAudioThreadProc(void *ctx) {
-    AAudioEngine *eng = reinterpret_cast<AAudioEngine *>(ctx);
-
-    bool status = TunePlayerForLowLatency(engine.playStream_);
-    if (!status) {
-        // if tune up is failed, audio could still play
-        LOGW("Failed to tune up the audio buffer size, low latency audio may not be guaranteed");
+aaudio_data_callback_result_t dataCallback(
+        AAudioStream *stream,
+        void *userData,
+        void *audioData,
+        int32_t numFrames) {
+
+    assert(userData && audioData);
+    AAudioEngine *eng = reinterpret_cast<AAudioEngine *>(userData);
+    assert(stream == eng->playStream_);
+
+    // Tuning the buffer size for low latency...
+    int32_t underRun = AAudioStream_getXRunCount(eng->playStream_);
+    if (underRun > eng->underRunCount_) {
+        /*
+         * Underrun happened since last callback:
+         * try to increase the buffer size.
+         */
+        eng->underRunCount_ = underRun;
+
+        aaudio_result_t actSize = AAudioStream_setBufferSizeInFrames(
+                stream, eng->bufSizeInFrames_ + eng->framesPerBurst_);
+        if (actSize > 0) {
+            eng->bufSizeInFrames_ = actSize;
+        } else {
+            LOGE("[AAudioManager] Output stream buffer tuning error: %s",
+                 AAudio_convertResultToText(actSize));
+        }
     }
-    // double check the tuning result: not necessary
-    PrintAudioStreamInfo(engine.playStream_);
 
-    LOGV("=====: currentState=%d", AAudioStream_getState(eng->playStream_));
+    int32_t samplesPerFrame = eng->sampleChannels_;
+    AUDIO_HARDWARE_SAMPLE_TYPE *bufferDest = static_cast<AUDIO_HARDWARE_SAMPLE_TYPE *>(audioData);
 
-    int32_t framesPerBurst = AAudioStream_getFramesPerBurst(eng->playStream_);
-    int32_t channelCount = AAudioStream_getSamplesPerFrame(eng->playStream_);
+    if (eng->playAudio_) {
+        PrintAudioStreamInfo(engine.playStream_);
 
-    LOGI("channel count %d", channelCount);
-    LOGI("framesPerBurst %d", framesPerBurst);
+        aaudio_stream_state_t aaudioStreamState = AAudioStream_getState(eng->playStream_);
+        LOGV("[AAudioManager] currentState=%d %s", aaudioStreamState,
+             AAudio_convertStreamStateToText(aaudioStreamState));
+//        if (aaudioStreamState == AAUDIO_STREAM_STATE_STARTING) {
+//            memset(static_cast<AUDIO_HARDWARE_SAMPLE_TYPE *>(audioData), 0,
+//                   sizeof(AUDIO_HARDWARE_SAMPLE_TYPE) * samplesPerFrame * numFrames);
+//            return AAUDIO_CALLBACK_RESULT_CONTINUE;
+//        }
```

Comment:
The AAudio miss misbehavior here: AAudio in `starting` state even after an async request... [Doc](https://developer.android.com/ndk/guides/audio/aaudio/aaudio.html)